### PR TITLE
Talk about exceptions to storing and updating headers

### LINKED ISSUE #165: Updating stored headers
7234 requires stored headers to be updated upon a `304` or `HEAD` response. 

However, there are a number of issues:

* It doesn't specify that H1 connection-related headers should be omitted
* Some browsers omit other headers (e.g., `x-content-*`, `content-*` and `x-webkit-*` - see [Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=174301) and [Webkit](https://bugs.webkit.org/show_bug.cgi?id=72414)
* Apache has a whitelist of headers; see [bug](https://bz.apache.org/bugzilla/show_bug.cgi?id=61820)
* Intermediaries are reluctant to do this for performance reasons; e.g., [squid](https://bugs.squid-cache.org/show_bug.cgi?id=7)

It seems like a lot of the bugs referenced above originated from the [confusing language around omitting entity headers in RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.5.3); that was removed in RFC7234 (see [current language](https://httpwg.org/specs/rfc7234.html#freshening.responses)).

I think we need to:

* [ ] Specify the list of H1 connection-oriented headers to be omitted
* [ ] Talk about other headers that need to be omitted (coordinating especially with browsers, who seem to have weird behaviour here)
* [ ] Contemplate any advice/warnings/requirements we can give to an intermediary cache about how they should handle this if they don't want the performance hit

[Mozilla's list of headers they don't update](https://dxr.mozilla.org/mozilla-central/source/netwerk/protocol/http/nsHttpResponseHead.cpp#879) might be a good starting point for the first two (see also [related bug](https://bz.apache.org/bugzilla/show_bug.cgi?id=61820)).

There are also [browser tests](https://wpt.fyi/results/fetch/http-cache/304-update.html?aligned&label=stable) and an [outstanding PR](https://github.com/web-platform-tests/wpt/pull/12316).

Note that this is a security issue, because some browsers filter out headers starting with `Content-`, which includes `Content-Security-Policy` -- i.e., a browser that has an old copy of the response won't see the new CSP header on a 304 response.

(this might also apply to 206 responses, since there's header combination there too)

### PR DESCRIPTION
**Pull request by mnot**
Fixes #165

### CODE REVIEW

**Comment by mnot** [comment]
Updated to use 'fields' as appropriate.

This test relies upon the understanding that `Connection` operates only on headers, not trailers; 2616 [seems to support that](https://tools.ietf.org/html/rfc2616#section-14.10). 

If that's the case, I think we need to make the appropriate changes in Messaging; @royfielding should we do that here, or in a separate issue first?

**Comment by royfielding** [comment]
The reason 

> The first MUST is right, but I believe the last three cases apply to both header and trailer fields. IOW, the names in Connection apply to trailers being forwarded as well.

The reason being that 2616 assumes the trailers would be merged as header fields (or simply dropped) prior to forwarding the message.

**Comment by mnot** [comment]
@royfielding that makes sense, although I suspect even fewer implementations are going to be doing that; might be worth reinforcing in Semantics?

In any case, please review.

File: draft-ietf-httpbis-cache-latest.xml [hunk 1]
```diff
@@ -386,6 +386,34 @@
...
+   before storage.</li>
+   <li>Likewise, fields that are known to be "hop-by-hop" are also required
+   to be removed before forwarding, and this &MAY; be implemented by doing so
+   before storage; see <xref target="header.connection"/> for some
+   examples.</li>
+   <li>Header fields that are specific to the connection they occur
+   &MUST-NOT; be stored. Effectively, this is limited to Proxy-Authenticate
+   <xref target="field.proxy-authenticate"/> and Proxy-Authorization <xref
+   target="field.proxy-authorization">, since new fields are
+   discouraged from being tied to a specific connection.</li>
```
Comment (mnot) [inline draft-ietf-httpbis-cache-latest.xml:409]: Note that this is dependent upon the resolution of #331, which may not go this way.

File: draft-ietf-httpbis-cache-latest.xml [hunk 2]
```diff
@@ -386,6 +386,34 @@
    prohibited from storing such responses.
 </t>
 
+<section title="Storing Header and Trailer Fields" anchor="storing.fields">
+<t>
+   Caches &MUST; include all received header fields &mdash; particularly
+   unrecognised ones &mdash; when storing a response; this assures that new HTTP
+   header fields can be successfully deployed. However, the following exceptions
```
Comment (reschke) [inline draft-ietf-httpbis-cache-latest.xml:393]: maybe "including" instead of "particularly"?

File: draft-ietf-httpbis-cache-latest.xml [hunk 3]
```diff
@@ -1043,8 +1071,14 @@
...
    to replace all instances of the corresponding header fields in the stored
-   response.
+   response, with the following exceptions:
 </t>
+<ul>
+   <li>The exceptions to header field storage in <xref
+   target="storing.fields"/> also apply to header field updates.</li>
+   <li>Caches &MUST-NOT; update the following header fields:
+   Content-Encoding, Content-Length, Content-MD5, Content-Range, ETag.</li>
+</ul>
```
Comment (reschke) [inline draft-ietf-httpbis-cache-latest.xml:1081]: Link the definitions? (I can update the PR)

File: draft-ietf-httpbis-cache-latest.xml (hunk 2 above)
Comment (mnot) [reply draft-ietf-httpbis-cache-latest.xml:393]: Done.

File: draft-ietf-httpbis-cache-latest.xml (hunk 3 above)
Comment (mnot) [reply draft-ietf-httpbis-cache-latest.xml:1081]: Please do!

Comment (reschke) [reply draft-ietf-httpbis-cache-latest.xml:1083]: Done.

**Comment by mnot** [comment]
OK, I think this is good to go with that last change.

@royfielding could you please review?

**Comment by mnot** [comment]
@royfielding see update

**Pull request merged**